### ISSUE TITLE
Use netlistsvg for schematic-like drawings

### ISSUE THREAD

**Issue by kasbah** [dvc94ch/pycircuit#14]
![](https://camo.githubusercontent.com/9b16db89aa566ab4b54e0b82f4ea82f6ebda0896/68747470733a2f2f63646e2e7261776769742e636f6d2f6e7475726c65792f6e65746c6973747376672f643031323835393436653035656535636539396463356430663830323562653538623539333661332f646f632f757033646f776e352e737667)

I just came across [Netlistsvg](https://github.com/nturley/netlistsvg) which uses [Klayjs](https://github.com/OpenKieler/klayjs) for rendering Yosys netlists as schematics. Looks like [Elkjs](https://github.com/OpenKieler/elkjs) is supposed to replace Klayjs. It's a JS port of the [Eclipse Layout Kernel](https://www.eclipse.org/elk). I haven't found any Python implementations yet. 

- [x] Output subset of Yosys netlist JSON that Netlistsvg supports
- [x] Make a Netlistsvg SVG skin for common analog components

**Comment by dvc94ch** [comment]
We don't need a python implementation. We can export json and modify netlistsvg to render it inside the viewer. We can probably add a export to svg or png feature inside the viewer. Or something along those lines. Thank you for the link!

**Comment by kasbah** [comment]
Yes, looking over it, outputting the subset of the Yosys netlist that Netlistsvg supports and adding some SVG "skins" for analog components could be a good way to go for a proof of concept. What do you mean by "the viewer" though?

An interactive viewer that lets you "zoom" into sub-circuits would be really cool. Worth noting from the Netlistsvg readme:

> Yosysjs and klayjs are both capable of handling sub circuits and subgraphs. It might be interesting to see what kind of images I can produce from a hierarchal design.

**Comment by dvc94ch** [comment]
> What do you mean by "the viewer" though

There is kind of a prototype in the viewer subfolder. You can start it with `node viewer/app.js net.svg pcb.svg` it's not very user friendly yet... pcb.svg doesn't have any styling built in it's in `viewer/css/pcb.svg.css`, the net highlighting is currently broken...

**Comment by kasbah** [comment]
Ah, completely missed that. Very cool! The style for the Graphviz output makes more sense now.

**Comment by kasbah** [comment]
I have started working on this as https://github.com/nturley/netlistsvg/pull/6/.

**Comment by dvc94ch** [comment]
Started work on this, like the results so far. this depends on #4, since port directions need to be specified and checked for consistency. This also allows removing the component labels, since they are no longer needed, new `Inst('R', '10K')` and `SubInst(joule_thief())`.

**Comment by dvc94ch** [comment]
Still pretty awful, using netlistsvg master. Correctly and consistently setting input and output information should make them look much better.

Power to signal or power to power nets should use vertical components, while signal to signal nets should use horizontal components. (analysis mostly implemented)

Port swapping and better skins should also make it look much better.

![screenshot from 2018-01-08 21-52-09](https://user-images.githubusercontent.com/741807/34691866-52491166-f4be-11e7-9513-6c7dbc04e02d.png)
![screenshot from 2018-01-08 21-52-16](https://user-images.githubusercontent.com/741807/34691869-5294ab30-f4be-11e7-8685-a80c5092651b.png)
![screenshot from 2018-01-08 21-52-23](https://user-images.githubusercontent.com/741807/34691870-52c7a634-f4be-11e7-9e9b-68d9485be9ca.png)

**Comment by kasbah** [comment]
You should add these examples to netlistsvg and open issues for features you mentioned that haven't been opened yet.

**Comment by nturley** [comment]
Hi, @dvc94ch . You raise some good points.

> Port swapping and better skins should also make it look much better.

Port swapping is definitely a problem. I'm still trying to figure out the rules for when to port swap. Better skins meaning more components?

>Power to signal or power to power nets should use vertical components, while signal to signal nets should use horizontal components

Right now I've been using a vertical flow for the analog skin, which does a bad job with horizontal ports so I've been just trying to use as many vertical components as possible, but presumably a typical circuit has more signal nets than power/gnd nets so maybe we should switch to a horizontal flow. Now that I have the lateral edge thing figured out, I might be able to make it look good. I'll try some experiments and see how it looks.

> The generic component should allow connecting nets at the top and bottom too

A simple fix would be to add some special info in the netlist for generic components, but I just tagged all the lateral ports in the skin, so maybe I should just start tagging all of the port directions in the netlist and use that to determine which ports are lateral.

Example 1:
----
![untitled](https://user-images.githubusercontent.com/2446659/34729044-3b123398-f521-11e7-9908-29371a6326f4.png)
Laterals branch fixed this problem. It just merged to master. Wires to horizontal ports were always messy until that merged.

---------

![untitled](https://user-images.githubusercontent.com/2446659/34729136-7867454e-f521-11e7-8f53-6f4e3b02b3b5.png)
This needs a port swap. The resistor is connected from a southward port to a lateral port so we should be able to determine the optimal direction. The fact that the southward port is VCC should be an even bigger hint about the correct resistor orientation.

-------

![untitled](https://user-images.githubusercontent.com/2446659/34729399-2c99d50e-f522-11e7-853f-4b504e31b04b.png)
Laterals branch does a better job with this one as well.

Example 2:
----------
![untitled](https://user-images.githubusercontent.com/2446659/34729826-5672bade-f523-11e7-9763-ebedad5873dc.png)

Laterals branch fixed this

-----------
![untitled](https://user-images.githubusercontent.com/2446659/34729673-fb0c5dd0-f522-11e7-8a4d-a948e1728be1.png)

Turn off LONGEST_PATH. I think I turned it back off on master. I still think that NETWORK_SIMPLEX is better for analog.

Example 3:
------------
![untitled](https://user-images.githubusercontent.com/2446659/34729968-acacdde4-f523-11e7-9188-f48554ff0992.png)
port swap, port swap, port swap.

So I think the issues for these layouts
* Port swap specific components (Especially when they are connected to VCC or GND) https://github.com/nturley/netlistsvg/issues/17
* See if we can get a better layout with a horizontal flow https://github.com/nturley/netlistsvg/issues/31
* Add more components in the skin (I don't think we need an issue for this because this will never be complete)
* Generics should have a way to get vertical ports https://github.com/nturley/netlistsvg/issues/30

**Comment by dvc94ch** [comment]
Progress so far:
Things that can still be done from the pycircuit side are the following:
- [ ] custom logic for port swapping transformer - needs to add dots to symbol to show current flow (joule_thief)
- [ ] improve finding circuit paths on signal nets (sallen key without psu example can't determine that the resistor needs port swapping)
- [ ] improve finding circuit paths for circuits without external ports (sallen key with psu connected)

![screenshot from 2018-01-10 17-30-25](https://user-images.githubusercontent.com/741807/34783540-2045a5de-f62c-11e7-9301-780991d8a9ec.png)
![screenshot from 2018-01-10 17-30-30](https://user-images.githubusercontent.com/741807/34783541-2075af9a-f62c-11e7-8c71-e62a0e3da31d.png)
![screenshot from 2018-01-10 17-30-43](https://user-images.githubusercontent.com/741807/34783542-208d1798-f62c-11e7-89f7-410a7af3ded7.png)
![screenshot from 2018-01-10 17-30-54](https://user-images.githubusercontent.com/741807/34783543-20adf81e-f62c-11e7-8218-301bfbb966b9.png)
![screenshot from 2018-01-10 17-31-03](https://user-images.githubusercontent.com/741807/34783544-20c66b38-f62c-11e7-84bb-beb2d5b8cfa2.png)
![screenshot from 2018-01-10 17-31-17](https://user-images.githubusercontent.com/741807/34783546-20e9724a-f62c-11e7-8816-7332380f5bbb.png)
